File: python/measure.py
import numpy as np
from numpy.random import default_rng
from typing import List

def measure_state_all(state, num_qubits, rng = None):
    # 计算测量概率
    probabilities = np.abs(state)**2
    # 生成随机数
    if rng == None:
        rng = default_rng()
    rand_num = rng.random()
    # 执行测量
    cumulative_prob = 0.0
    for idx, prob in enumerate(probabilities):
        cumulative_prob += prob
        if rand_num < cumulative_prob:
            # measured_state = np.zeros_like(state)
            # measured_state[idx] = 1.0
            measured_state = format(idx, f'0{num_qubits}b')
            return measured_state

def measure_sample_all(final_state, num_qubits: int, shot:int = 1024, seed = None):
    rng = default_rng(seed=seed)  # 使用默认的随机数生成器，并设置随机种子
    counts = {}
    for _ in range(shot):
        m = measure_state_all(final_state, num_qubits, rng)
        counts[m] = counts.get(m,0) + 1
    return counts

# 局部测量
def measure_sample(final_state, num_qubits: int, qubits_to_measure:List[int], shot:int = 1024, seed = None):
    dimension = 2**num_qubits
    
    num_qubits_to_measure = len(qubits_to_measure)
    
    # 生成测量矩阵
    measurement_matrix = np.zeros((2**num_qubits_to_measure, dimension))
    for i in range(2**num_qubits_to_measure):
        basis_state = np.binary_repr(i, width=num_qubits_to_measure)
        basis_index = sum([int(b) * 2**(num_qubits_to_measure - j - 1) for j, b in enumerate(basis_state)])
        for j, qubit in enumerate(qubits_to_measure):
            basis_index += int(basis_state[j]) * 2**qubit
        measurement_matrix[i, basis_index] = 1.0
    
    # 得到局部的测量状态
    measured_state = measurement_matrix @ final_state
    
    # 归一化测量态
    normalized_measured_state = measured_state / np.linalg.norm(measured_state)
    print("measured_state:", measured_state)
    probability_sum = np.sum(np.abs(measured_state)**2)
    print("Total probability after measurement:", probability_sum)
    counts = measure_sample_all(normalized_measured_state, num_qubits_to_measure, shot, seed)
    
    return counts

File: rust/src/measure.rs
use ndarray::ArrayBase;
use ndarray::OwnedRepr;
use ndarray::Dim;
use rand::prelude::*;
use std::collections::HashMap;

/// 计算测量概率并返回测量结果的二进制表示
pub fn measure_state_all(state: &ArrayBase<OwnedRepr<num_complex::Complex<f64>>, Dim<[usize; 1]>>, num_qubits: usize, rng: &mut dyn RngCore) -> String {
    // 计算测量概率
    let probabilities: Vec<f64> = state.iter().map(|&amplitude| amplitude.norm_sqr()).collect();
    let rand_num: f64 = rng.gen();

    let mut cumulative_prob = 0.0;
    for (idx, &prob) in probabilities.iter().enumerate() {
        cumulative_prob += prob;
        if rand_num < cumulative_prob {
            return format!("{:0width$b}", idx, width = num_qubits);
        }
    }

    "0".repeat(num_qubits)
}

/// 对最终量子态进行多次采样测量
pub fn measure_sample_all(final_state: &ArrayBase<OwnedRepr<num_complex::Complex<f64>>, Dim<[usize; 1]>>, num_qubits: usize, shot: usize, seed: u64) -> HashMap<String, usize> {
    let mut rng = StdRng::seed_from_u64(seed);
    let mut counts = HashMap::new();

    for _ in 0..shot {
        let m = measure_state_all(final_state, num_qubits, &mut rng);
        *counts.entry(m).or_insert(0) += 1;
    }

    counts
}
